internal/utils/validation.go

```diff
@@ -8,95 +8,129 @@ SPDX-License-Identifier: Apache-2.0
+				continue
+			}
+			for _, qr := range rqi.GetRawReads().KvReads {
+				k := TransformToFPCKey(qr.Key)
+				v, err := stub.GetState(k)
+				if err != nil {
+					return fmt.Errorf("error (%s) reading key %s", err, k)
+				}
+
+				_ = v
+				return fmt.Errorf("TODO: not implemented, missing hash check")
+			}
 		}
 	}
 
 	// writes
-	var writeKeys []string
-	writesetMap := make(map[string][]byte)
-	for _, w := range rwset.Writes {
-		k := TransformToFPCKey(w.Key)
-		writeKeys = append(writeKeys, k)
-		writesetMap[k] = w.Value
-		_ = stub.PutState(k, w.Value)
-	}
-
-	// sort readset and writeset as enclave uses a sorted map
-	sort.Strings(readKeys)
-	sort.Strings(writeKeys)
-
-	// prepare sorted read/write set as output
-	for _, k := range readKeys {
-		readset = append(readset, []byte(k))
-		readset = append(readset, readsetMap[k])
-	}
-
-	for _, k := range writeKeys {
-		writeset = append(writeset, []byte(k))
-		writeset = append(writeset, writesetMap[k])
+	if rwset.GetWrites() != nil {
+		logger.Debugf("Replaying writes")
+		for _, w := range rwset.Writes {
+			k := TransformToFPCKey(w.Key)
+			err := stub.PutState(k, w.Value)
+			if err != nil {
+				return fmt.Errorf("error (%s) writing key %s value(hex) %s", err, k, hex.EncodeToString(w.Value))
+			}
+
+			logger.Debugf("written key %s value(hex) %s", k, hex.EncodeToString(w.Value))
+		}
 	}
 
-	return readset, writeset, nil
+	return nil
 }
 
-func Validate(responseMsg *protos.ChaincodeResponseMessage, readset, writeset [][]byte, attestedData *protos.AttestedData) error {
-	if responseMsg.Signature == nil {
+func Validate(signedResponseMessage *protos.SignedChaincodeResponseMessage, attestedData *protos.AttestedData) error {
+	if signedResponseMessage.Signature == nil {
 		return fmt.Errorf("absent enclave signature")
 	}
 
-	// Note: below signature was created in ecc_enclave/enclave/enclave.cpp::gen_response
-	hash := ComputedHash(responseMsg, readset, writeset)
+	// prepare and do signature verification
+	enclaveVkPtr := C.CBytes(attestedData.EnclaveVk)
+	defer C.free(enclaveVkPtr)
 
-	// perform enclave signature validation
-	// TODO refactor this to function
-	pub, err := x509.ParsePKIXPublicKey(attestedData.EnclaveVk)
-	if err != nil {
-		return err
-	}
+	responseMessagePtr := C.CBytes(signedResponseMessage.ChaincodeResponseMessage)
+	defer C.free(responseMessagePtr)
+
+	signaturePtr := C.CBytes(signedResponseMessage.Signature)
+	defer C.free(signaturePtr)
 
-	valid := ecdsa.VerifyASN1(pub.(*ecdsa.PublicKey), hash[:], responseMsg.Signature)
-	if !valid {
-		return fmt.Errorf("signature invalid")
+	ret := C.verify_signature((*C.uint8_t)(enclaveVkPtr), C.uint32_t(len(attestedData.EnclaveVk)), (*C.uint8_t)(responseMessagePtr), C.uint32_t(len(signedResponseMessage.ChaincodeResponseMessage)), (*C.uint8_t)(signaturePtr), C.uint32_t(len(signedResponseMessage.Signature)))
+	if ret == false {
+		return fmt.Errorf("enclave signature verification failed")
 	}
```

Comment:
Hmm, would be nicer to abstract into a separate go function, e.g., in `internal/utils/crypto.go` (where now also the `ComputeHash` function is obsolete).  You will need this function also in mock and it could centralize all corresponding cgo stuff (and eventually also encryption) in a single and shareble file?

Reply:
There is a lot of refactoring to be done. I have opened 501 to track the tasks, and maybe proceed in parallel